File: windows/Wetterkurve.Core/WeatherModels.cs
using System.Text.Json.Serialization;

namespace Wetterkurve;

public sealed record WeatherLocation(
    string Id,
    string Name,
    string Label,
    double Latitude,
    double Longitude,
    string Timezone);

public sealed record HourlyPoint(
    DateTime Time,
    double Temperature,
    double Apparent,
    double PrecipitationProbability,
    double Precipitation,
    int WeatherCode,
    double Wind);

public sealed record CurrentWeather(
    [property: JsonPropertyName("temperature_2m")] double Temperature,
    [property: JsonPropertyName("apparent_temperature")] double ApparentTemperature,
    [property: JsonPropertyName("weather_code")] int WeatherCode,
    [property: JsonPropertyName("wind_speed_10m")] double WindSpeed,
    [property: JsonPropertyName("relative_humidity_2m")] double RelativeHumidity);

public sealed record HourlyWeather(
    [property: JsonPropertyName("time")] List<string> Time,
    [property: JsonPropertyName("temperature_2m")] List<double> Temperature,
    [property: JsonPropertyName("apparent_temperature")] List<double> ApparentTemperature,
    [property: JsonPropertyName("precipitation_probability")] List<double> PrecipitationProbability,
    [property: JsonPropertyName("precipitation")] List<double> Precipitation,
    [property: JsonPropertyName("weather_code")] List<int> WeatherCode,
    [property: JsonPropertyName("wind_speed_10m")] List<double> WindSpeed);

public sealed record ForecastPayload(
    [property: JsonPropertyName("current")] CurrentWeather Current,
    [property: JsonPropertyName("hourly")] HourlyWeather Hourly);

public sealed record GeocodingHit(
    [property: JsonPropertyName("name")] string? Name,
    [property: JsonPropertyName("latitude")] double Latitude,
    [property: JsonPropertyName("longitude")] double Longitude,
    [property: JsonPropertyName("admin1")] string? Admin1,
    [property: JsonPropertyName("country")] string? Country,
    [property: JsonPropertyName("timezone")] string? Timezone);

public sealed record GeocodingResponse(
    [property: JsonPropertyName("results")] List<GeocodingHit>? Results);

public sealed record DaySegment(string Key, string Label, int Start, int End);

public sealed record WeatherCondition(string Symbol, string Description, string Icon);

public sealed record AppState
{
    public List<WeatherLocation> Locations { get; set; } = [];
    public int ActiveLocation { get; set; }
    public double? IndicatorLeft { get; set; }
    public double? IndicatorTop { get; set; }
}

File: android/core/src/main/kotlin/de/wean/wetterkurve/WeatherModels.kt
package de.wean.wetterkurve

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import java.time.LocalDateTime

data class WeatherLocation(
    val id: String,
    val name: String,
    val label: String,
    val latitude: Double,
    val longitude: Double,
    val timezone: String,
)

data class HourlyPoint(
    val time: LocalDateTime,
    val temperature: Double,
    val apparent: Double,
    val precipitationProbability: Double,
    val precipitation: Double,
    val weatherCode: Int,
    val wind: Double,
    val cloudCover: Double = 0.0,
)

data class DaySegment(
    val key: String,
    val label: String,
    val start: Int,
    val end: Int,
)

data class WeatherCondition(
    val symbol: String,
    val description: String,
    val icon: String,
)

data class AppState(
    val locations: List<WeatherLocation>,
    val activeLocation: Int,
    val showClouds: Boolean = true,
    val showWind: Boolean = true,
)

@Serializable
data class ForecastPayload(
    val current: CurrentWeather? = null,
    val hourly: HourlyWeather? = null,
)

@Serializable
data class CurrentWeather(
    @SerialName("temperature_2m") val temperature: Double,
    @SerialName("apparent_temperature") val apparentTemperature: Double,
    @SerialName("weather_code") val weatherCode: Int,
    @SerialName("wind_speed_10m") val windSpeed: Double,
    @SerialName("relative_humidity_2m") val relativeHumidity: Double,
)

@Serializable
data class HourlyWeather(
    val time: List<String> = emptyList(),
    @SerialName("temperature_2m") val temperature: List<Double> = emptyList(),
    @SerialName("apparent_temperature") val apparentTemperature: List<Double> = emptyList(),
    @SerialName("precipitation_probability") val precipitationProbability: List<Double> = emptyList(),
    @SerialName("precipitation") val precipitation: List<Double> = emptyList(),
    @SerialName("weather_code") val weatherCode: List<Int> = emptyList(),
    @SerialName("wind_speed_10m") val windSpeed: List<Double> = emptyList(),
    @SerialName("cloud_cover") val cloudCover: List<Double> = emptyList(),
)

@Serializable
data class GeocodingHit(
    val name: String? = null,
    val latitude: Double = Double.NaN,
    val longitude: Double = Double.NaN,
    val admin1: String? = null,
    val country: String? = null,
    val timezone: String? = null,
)

@Serializable
data class GeocodingResponse(
    val results: List<GeocodingHit>? = null,
)

@Serializable
data class WeatherLocationJson(
    val id: String? = null,
    val name: String? = null,
    val label: String? = null,
    val latitude: Double = Double.NaN,
    val longitude: Double = Double.NaN,
    val timezone: String? = null,
)

@Serializable
data class AppStateJson(
    val locations: List<WeatherLocationJson> = emptyList(),
    val activeLocation: Int = 0,
    val showClouds: Boolean = true,
    val showWind: Boolean = true,
)
